fastlane/fastlane
# [WIP][spaceship] add initial App Clips support

### Checklist

### Motivation and Context



https://github.com/fastlane/fastlane/issues/17639

We got no App Clips support in the official App Store Connect API from WWDC21 so I came to start supporting it with the web session one. 

### Description



I have added the following four new models to support the basic settings of App Clips' metadata on App Store Connect console, which represents the information on the screenshot.

* appClips - this will be automatically created once you upload an ipa file including an App Clip
* appClipVersions - includes Action
* appClipVersionLocalizations - includes locale and subtitle
* appClipHeaderImages - image

<img width="885" alt="Screenshot 2021-07-04 at 00 46 57" src="https://user-images.githubusercontent.com/748949/124360891-38f11d00-dc67-11eb-873c-637eb4795969.png">

JFYI: I haven't looked into things that can be edited within "Edit Advanced Experiences" pages.

### Testing Steps



<img width="883" alt="Screenshot 2021-07-04 at 11 36 23" src="https://user-images.githubusercontent.com/748949/124371361-11786f80-dcbc-11eb-8dd8-47726c1205af.png">

As the above says, to see the App Clip section enabled on App Store Connect, you would need to follow these. 

* Create an app including App Clips - This is a good example from Apple https://developer.apple.com/documentation/swiftui/fruta_building_a_feature-rich_app_with_swiftui
* Upload the binary via deliver

Then you'll see the section enabled, which means `appClips` and `appClipVersions` resources are created internally already. You can play APIs with the spaceship playground command, like below.

Note that this currently works with only the web session. App Store Connect API key wont' work for this.

```ruby
app = Spaceship::ConnectAPI::App.find('YOUR_APP_BUNDLE_ID')
app_clip = app.get_app_clips.first
app_clip_version = app_clip.get_app_clip_versions.first

app_clip_version.update(attributes: { invocationVerb: Spaceship::ConnectAPI::AppClipVersion::InvocationVerb::OPEN })
app_clip_version.create_app_clip_version_localization(attributes: { locale: 'ja', subtitle: 'サブタイトル' } )

localizations = app_clip_version.get_app_clip_version_localizations
localizations[0].update(attributes: { subtitle: 'updated subtitle!' })

app_clip_header_image = Spaceship::ConnectAPI::AppClipHeaderImage.create(app_clip_version_localization_id: localizations[0].id, path: 'path/to/image')
app_clip_header_image.delete!
```

FYI: Apparently, once you change things within "Edit Adavanced Experiences", you won't be able to modify some attributes via `AppClipVersion#update` 🤔 

> [4] pry(#<Spaceship::Playground>)> app_clip_version.update(attributes: { invocationVerb: Spaceship::ConnectAPI::AppClipVersion::InvocationVerb::VIEW })
> Spaceship::UnexpectedResponse: An attribute value is not acceptable for the current resource state. - The attribute 'invocationVerb' cannot be modified at this time. - /data/attributes/invocationVerb
